src/System.Runtime.Numerics/src/System/Numerics/BigIntegerCalculator.Buffer.cs

```diff
@@ -0,0 +1,123 @@
+
+        internal struct BitsBuffer
+        {
+            private uint[] _bits;
+            private int _length;
+
+            public BitsBuffer(int length, uint value)
+            {
+                Debug.Assert(length >= 1);
+
+                _bits = new uint[length];
+                _length = value != 0 ? 1 : 0;
+
+                _bits[0] = value;
+            }
+
+            public BitsBuffer(int length, uint[] value)
+            {
+                Debug.Assert(value != null);
+                Debug.Assert(length >= ActualLength(value));
+
+                _bits = new uint[length];
+                _length = ActualLength(value);
+
+                Array.Copy(value, _bits, _length);
+            }
+
+            [SecuritySafeCritical]
+            public unsafe void MultiplySelf(ref BitsBuffer value,
+                                            ref BitsBuffer temp)
+            {
+                Debug.Assert(temp._length == 0);
+                Debug.Assert(_length + value._length <= temp._bits.Length);
+
+                fixed (uint* b = _bits, v = value._bits, t = temp._bits)
+                {
+                    if (_length < value._length)
+                    {
+                        Multiply(v, value._length,
+                                 b, _length,
+                                 t, _length + value._length);
+                    }
+                    else
+                    {
+                        Multiply(b, _length,
+                                 v, value._length,
+                                 t, _length + value._length);
+                    }
+                }
+
+                Apply(ref temp, _length + value._length);
+            }
+
+            [SecuritySafeCritical]
+            public unsafe void SquareSelf(ref BitsBuffer temp)
+            {
+                Debug.Assert(temp._length == 0);
+                Debug.Assert(_length + _length <= temp._bits.Length);
+
+                fixed (uint* b = _bits, t = temp._bits)
+                {
+                    Square(b, _length,
+                           t, _length + _length);
+                }
+
+                Apply(ref temp, _length + _length);
+            }
+
+            public void Reduce(ref FastReducer reducer)
+            {
+                _length = reducer.Reduce(_bits, _length);
+            }
+
+            [SecuritySafeCritical]
+            public unsafe void Reduce(uint[] modulus)
+            {
+                Debug.Assert(modulus != null);
+
+                if (_length >= modulus.Length)
+                {
+                    fixed (uint* b = _bits, m = modulus)
+                    {
+                        Divide(b, _length,
+                               m, modulus.Length,
+                               null, 0);
+                    }
+
+                    _length = ActualLength(_bits, modulus.Length);
+                }
+            }
+
+            public uint[] ToArray()
```

Comment:
This is all internal so it's not as big a deal, but could you rename this?  Typically ToArray has connotations of creating a new copy, whereas this is just returning the underlying storage that could still be in use.  Maybe GetBuffer like MemoryStream uses?